hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/CleanerChore.java

```diff
@@ -322,140 +330,171 @@ public boolean getEnabled() {
+   * @param dir means the directory we will start to traverse and delete.
+   * @param root means whether it's the root directory to traverse, if true then cannot delete it.
+   * @param result {@link AsyncResult<Boolean>} to fetch the result. True means the current
+   *          directory has been deleted successfully (for root dir we don't need that) and the
+   *          parent will try to delete its own directory if all of the children(files and
+   *          sub-directories are included) has been deleted successfully.
+   */
+  private void traverseAndDelete(final Path dir, final boolean root,
+      final AsyncResult<Boolean> result) {
+    try {
+      final Action<Boolean> curDirDeletion = new Action<Boolean>() {
+        @Override
+        public Boolean act() throws IOException {
+          return fs.delete(dir, false);
         }
-        allSubdirsDeleted = deleteAction(new Action<Boolean>() {
-          @Override
-          public Boolean act() throws IOException {
-            return getCleanResult(tasks);
-          }
-        }, "subdirs");
+      };
+
+      // Step.1: List all files under the given directory.
+      List<FileStatus> allPaths = Arrays.asList(fs.listStatus(dir));
+      final List<FileStatus> subDirs = new ArrayList<>();
+      final List<FileStatus> files = new ArrayList<>();
+      for (FileStatus status : allPaths) {
+        if (status.isDirectory()) {
+          subDirs.add(status);
+        } else if (status.isFile()) {
+          files.add(status);
+        }
+      }
+
+      // Step.2: Try to delete all the deletable files.
+      final boolean allFilesDeleted = files.isEmpty() || deleteAction(new Action<Boolean>() {
+        @Override
+        public Boolean act() throws IOException {
+          return checkAndDeleteFiles(files);
+        }
+      }, "files", dir);
+
+      // Step.3: Start to traverse and delete the sub-directories.
+      if (subDirs.isEmpty()) {
+        // If no sub-directories, then just try to delete the current dir and finish the result.
+        boolean deleted = allFilesDeleted;
+        if (allFilesDeleted && !root) {
+          deleted = deleteAction(curDirDeletion, "dir", dir);
+        }
+        result.set(deleted);
+        return;
       }
 
-      boolean result = allFilesDeleted && allSubdirsDeleted;
-      // if and only if files and subdirs under current dir are deleted successfully, and
-      // it is not the root dir, then task will try to delete it.
-      if (result && !root) {
-        result &= deleteAction(new Action<Boolean>() {
+      // Otherwise, there should be some sub-directories. then we will register the following
+      // callback in AsyncResult of sub-directory, and once all of the sub-directories are traversed
+      // and deleted then the callback will try to delete the current dir and finish the result.
+      final AtomicInteger remain = new AtomicInteger(subDirs.size());
+      Callback<Boolean> callback = new Callback<Boolean>() {
```

Comment:
Where does this `callback.get()` get called? Or is it no need?  Asking for sure.

Reply:
You mean where will we call the callback#run ?  You can see the AsyncResult, once the uppler layer set the result to AsyncResult, then it will call the callback#run.  For now, the callback is used for cleaning the current dir **asynchronsely** once all sub-dirs & sub-files got deleted.
Thanks.